### PR TITLE
rust: Write config.toml not config

### PR DESCRIPTION
**Pull request by mmlb**
## Description of changes

Changes the rust infra to write cargo config to .cargo/config.toml insted of .cargo/config since the latter is deprecated (since 1.38?) and is now warning on stderr.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by mmlb** [comment]
Tried running nixpkgs-review but that failed with merge conflicts :/

**Review by AndersonTorres — APPROVED** [review]

**Review by drupol — APPROVED** [review]
LGTM at first sight.

**Comment by mweinelt** [comment]
It feels like this change caused cargoDeps hashes to be invalidated. Why was this merged without a review from a codeowner?

**Comment by drupol** [comment]
My bad, should I provide a revert ?

**Comment by mweinelt** [comment]
For codeowners to decide.

**Comment by winterqt** [comment]
Yes, this was irresponsible to merge in its current state, and I'm sorry I didn't see it in my inbox.

For now, let's just revert the whole thing to avoid any breakages making it to master.

**Comment by drupol** [comment]
My apologies for the extra work.

**Comment by winterqt** [comment]
We should be able to pick all the changes except the one to `fetchCargoTarball`.

**Comment by winterqt** [comment]
*Any* change to the output of the FODs will require hash changes, which is why I said we should first just change everything but the FODs.

**Comment by mmlb** [comment]
Bummer, I would have expected ofborg or something to catch this. I would actually prefer dealing with the hash change instead of layering fixes/hacks. Especially if upstream rust is doing this then it would be easier to avoid papering over things like this.

**Comment by alyssais** [comment]
Nix FODs aren't really designed for hashes to ever change — even if we fix all the hashes in Nixpkgs, we silently break all out of tree FODs, which means most out of tree Cargo hashes.  Justifying a FOD hash change is an _extremely_ high bar — there's a reason we have fetchpatch2.  (Generally introducing a new API with new hashes like that has fewer negative consequences.)

**Comment by mmlb** [comment]
Hmm I'm confused about the revert though, I didn't have any need to update the cargo hash in commitmsgmft and had successfull build/test there.

**Comment by winterqt** [comment]
Did you try busting the hash with `lib.fakeHash` to ensure the hashes were the same before and after?

**Comment by alyssais** [comment]
> Hmm I'm confused about the revert though, I didn't have any need to update the cargo hash in commitmsgmft and had successfull build/test there.

This is exactly why FOD hash changes are so problematic — Nix won't ever rebuild a FOD it's built before, so you don't get to find out your FODs are broken for other people unless they tell you.

**Comment by mmlb** [comment]
> Did you try busting the hash with `lib.fakeHash` to ensure the hashes were the same before and after?

Ah nope.

> > Hmm I'm confused about the revert though, I didn't have any need to update the cargo hash in commitmsgmft and had successfull build/test there.
> 
> This is exactly why FOD hash changes are so problematic — Nix won't ever rebuild a FOD it's built before, so you don't get to find out your FODs are broken for other people unless they tell you.

Yep it makes sense now. I was only having issues with commitmsgmft tests.

I did end up rebuilding icedtea-web iirc and that ended up rebuilding Firefox and a bunch of other stuff, but I suppose that is the same situation...

File: pkgs/build-support/rust/hooks/cargo-setup-hook.sh
Comment (mmlb) [reply pkgs/build-support/rust/hooks/cargo-setup-hook.sh:25]: I'll give this a try.

**Comment by mmlb** [comment]
I'll give this another go withess disruptive changes then :D

**Comment by AndersonTorres** [comment]
> there's a reason we have fetchpatch2

Does it apply here? A "fetchCargo2"?

**Comment by alyssais** [comment]
We could, although I'm not sure it's worth it just for this…

File: pkgs/build-support/rust/hooks/cargo-setup-hook.sh
Comment (mmlb) [reply pkgs/build-support/rust/hooks/cargo-setup-hook.sh:25]: Actually this wasn't the problem, the patch that caused FOD changes was the `fetch-cargo-tarball` one!

**Pull request merged**
